CMakeLists.txt

```diff
@@ -416,86 +263,118 @@ if (S2N_INTERN_LIBCRYPTO)
-        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/ktls.c"
-)
-if(PLATFORM_SUPPORTS_KTLS)
-    message(STATUS "feature probe: kTLS supported")
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_PLATFORM_SUPPORTS_KTLS)
-endif()
+# Iterate over all of the features and try to compile them
+FILE(GLOB FEATURE_SRCS "${CMAKE_CURRENT_LIST_DIR}/tests/features/*.c")
+list(SORT FEATURE_SRCS)
+foreach(file ${FEATURE_SRCS})
+    get_filename_component(feature_name ${file} NAME_WE)
+    feature_probe(${feature_name})
+endforeach()
 
-if (NOT DEFINED CMAKE_AR)
-	message(STATUS "CMAKE_AR undefined, setting to `ar` by default")
-	SET(CMAKE_AR ar)
-else()
-	message(STATUS "CMAKE_AR found: ${CMAKE_AR}")
+# FreeBSD might need to link to execinfo explicitly
+if(NOT S2N_EXECINFO_AVAILABLE AND CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+    feature_probe(S2N_EXECINFO_AVAILABLE LINK_LIBRARIES execinfo)
 endif()
 
-if (NOT DEFINED CMAKE_RANLIB)
-    message(STATUS "CMAKE_RANLIB undefined, setting to `ranlib` by default")
-    SET(CMAKE_RANLIB ranlib)
-else()
-    message(STATUS "CMAKE_RANLIB found: ${CMAKE_RANLIB}")
+# Stack traces are only available if execinfo is
+if (NOT S2N_EXECINFO_AVAILABLE)
+    set(S2N_STACKTRACE FALSE)
 endif()
+feature_probe_result(S2N_STACKTRACE ${S2N_STACKTRACE})
 
-if (NOT DEFINED CMAKE_OBJCOPY)
-    message(STATUS "CMAKE_OBJCOPY undefined, setting to `objcopy` by default")
-    SET(CMAKE_OBJCOPY objcopy)
-else()
-    message(STATUS "CMAKE_OBJCOPY found: ${CMAKE_OBJCOPY}")
+set(S2N_KYBER512R3_AVX2_BMI2 FALSE)
+if(NOT S2N_NO_PQ_ASM)
+    # Kyber Round-3 code has several different optimizations which require
+    # specific compiler flags to be supported by the compiler.
+    # So for each needed instruction set extension we check if the compiler
+    # supports it and set proper compiler flags to be added later to the
+    # Kyber compilation units.
+    if(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^(x86_64|amd64|AMD64)$")
+        # Some platforms support -mavx2 flag but not m256 intrinsics required to use them. Only enable Kyber assembly
+        # optimizations if both are supported. See https://github.com/aws/s2n-tls/pull/3005 for more info.
+        if(S2N_KYBER512R3_AVX2_BMI2_SUPPORTED AND S2N_KYBER512R3_M256_INTRINSICS_SUPPORTED)
+            set(S2N_KYBER512R3_AVX2_BMI2 TRUE)
+            enable_language(ASM)
+
+            # add the assembly files to the project
+            FILE(GLOB KYBER512R3_AVX2_BMI2_ASM_SRCS "pq-crypto/kyber_r3/*_avx2.S")
+            target_sources(${PROJECT_NAME} PRIVATE ${KYBER512R3_AVX2_BMI2_ASM_SRCS})
+
+            # compile the C files with avx flags
+            FILE(GLOB KYBER512R3_AVX2_BMI2_SRCS "pq-crypto/kyber_r3/*_avx2.c")
+            set_source_files_properties(${KYBER512R3_AVX2_BMI2_SRCS} PROPERTIES COMPILE_FLAGS ${S2N_KYBER512R3_AVX2_BMI2_SUPPORTED_FLAGS})
+        endif()
+    endif()
 endif()
+feature_probe_result(S2N_KYBER512R3_AVX2_BMI2 ${S2N_KYBER512R3_AVX2_BMI2})
```

Comment:
Am I missing something, or is this only supported via cmake, not for the bindings or make? Is that a problem?

But if that's right, then the same problem for S2N_STACKTRACE seems more serious...